Wishlist Rest API implementation

Tickles-BV opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios,
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

This pull request is a small start at adding an api part for the wishlist functionality in magento 2.
Because it adds new interfaces I have tried to only implement the smallest possible set of changes so this can be discussed and implemented further.

I have added a wishlist repository, which allows one to retrieve a wishlist for a customer, and to add products to a customer's wishlist. I've also added an interface for the wishlist model itself, also for use in the rest api. The wishlist model interface currently only interfaces 'getName', because I'm unsure which properties should be added to it.

I have a few questions about implementation details.

1. How do I know which attributes should be made available in data api? For instance, should you get a list of wishlist products when retrieving a wishlist by customer? Or should you simply get a wishlist id, which can then be used to retrieve the items via another API call. This seems like a cleaner seperation of logic, but I'd like confirmation.
2. I've added two methods to the repository: `getWishlistForCustomer` and `addWishlistForCustomer`, should these be in the repository? Or should these be extracted to a WishlistManagement class.
3. I've simply made the api route 'V1/wishlist' are there conventions for this? If so, where can I find these?
4. In the wishlist model, method 'getSharingCode' is added via @property docblock tag, can I add these to api response without adding to new const and wrapping in own method?

 Sorry for the amount of questions but I'm trying to wrap my head around a lot of magento 2 internals implementing this.


### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/community-features#56: API: Wishlist


### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Tickles-BV committed:
Add api routes for wishlist repository

Tickles-BV committed:
Add preference for WishlistRepositoryInterface to dependency injector

Tickles-BV committed:
Create WishlistInterface and add to wishlist model

Tickles-BV committed:
Add wishlistrepository and interface

Tickles-BV committed:
Re-add getSharingCode property

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=16075) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=16075) before we can accept your contribution.<br/><hr/>**Wietze Mink** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

magento-engcom-team commented:
Hi @Tickles-BV. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on Pull Request changes
- `@magento-engcom-team give me new test instance` - deploy **NEW** test instance based on Pull Request changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy Vanilla Magento instance for Issue or Pull Request

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

Tickles-BV committed:
Add CONSTs and getters and setters for data items in wishlist model

Tickles-BV committed:
Implement interface and remove magic @method docblocks, adding concrete ones instead

Tickles-BV committed:
Add docblocks and small stuff to wishlistinterface

Tickles-BV commented:
I've updated the WishlistInterface and now it encapsulates all the properties available in the wishlist model, wishlist model now properly implements this and all the data is returned in api calls. 

This answers both question 1 and 4 on my end, the others I'd still like to have someone look at.

sidolov commented on app/code/Magento/Wishlist/Api/WishlistRepositoryInterface.php:
```diff
@@ -0,0 +1,26 @@
+ * @api
+ * @package Magento\Wishlist\Api
+ */
+interface WishlistRepositoryInterface
```
1. The repository should have methods for CRUD operations: get, getList, save and delete. 
2. Wishlist Repository should manipulate Wishlist entity only 
3. Use strict types for methods

sidolov commented on app/code/Magento/Wishlist/Api/data/WishlistInterface.php:
```diff
@@ -0,0 +1,87 @@
+<?php
+
+namespace Magento\Wishlist\Api\data;
```
First letter in namespace should be uppercase - `Magento\Wishlist\Api\Data`

Tickles-BV replied:
Resolved

sidolov commented on app/code/Magento/Wishlist/Api/data/WishlistInterface.php:
```diff
@@ -0,0 +1,87 @@
+
+    const SHARED = 'shared';
+
+    const UPDATED_AT = 'updated_at';
```
This field seems to be redundant. Do you have the specific case?

Tickles-BV replied:
Should not all fields used by getter and setter methods be specified in api constants?

sidolov commented on app/code/Magento/Wishlist/Api/data/WishlistInterface.php:
```diff
@@ -0,0 +1,87 @@
+    /**#@+
+     * Constants defined for keys of  data array
+     */
+    const NAME = 'name';
```
Wishlist name is immutable in the current implementation, why we need this field in api?

Tickles-BV replied:
Should api not have method to retrieve name? Or should I only remove NAME constant and keep getName method and remove setName method? If so, why remove constant?

sidolov commented on app/code/Magento/Wishlist/Api/data/WishlistInterface.php:
```diff
@@ -0,0 +1,87 @@
+
+namespace Magento\Wishlist\Api\data;
+
+interface WishlistInterface
```
1. Some setters for fields are missing 
2. Use strict types for methods 
3. Setter and getter for extension attributes are missing

Tickles-BV replied:
I've added setter for customerId, with 2, do you mean declare('strict_types') or that I should use typehints

sidolov commented on app/code/Magento/Wishlist/Model/Wishlist.php:
```diff
@@ -486,17 +490,15 @@ public function addNewItem($product, $buyRequest = null, $forciblySetQty = false
     public function setCustomerId($customerId)
     {
-        return $this->setData($this->_getResource()->getCustomerIdFieldName(), $customerId);
+        return $this->setData(self::CUSTOMER_ID, $customerId);
```
This change may break 3d party extension that uses this extension point for customer_id field

Tickles-BV replied:
This is the way it is implemented in new code such as BookmarkInterface, what is preferred solution? Removing field from constants and just keeping getId and setId in interface?

sidolov commented on app/code/Magento/Wishlist/Model/WishlistRepository.php:
```diff
@@ -0,0 +1,66 @@
+        $product = $this->productRepository->getById($productId);
+
+        try {
+            $item = $wishlist->addNewItem($product);
```
Wishlist item is the different entity and managing should be in a separate repository (for example WishlistItemRepository)

sidolov commented on app/code/Magento/Wishlist/Model/WishlistRepository.php:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * @inheritdoc
+     */
+    public function addWishlistForCustomer($customerId, $productId)
```
Please, try to use product SKU instead of product id.

sidolov commented on app/code/Magento/Wishlist/etc/di.xml:
```diff
@@ -8,6 +8,8 @@
 <config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
     <preference for="Magento\Wishlist\Model\AuthenticationStateInterface" type="Magento\Wishlist\Model\AuthenticationState" />
     <preference for="Magento\Wishlist\Controller\WishlistProviderInterface" type="Magento\Wishlist\Controller\WishlistProvider" />
+    <preference for="Magento\Wishlist\Api\WishlistRepositoryInterface" type="Magento\Wishlist\Model\WishlistRepository" />
```
Preference for WishlistInterface is missing

Tickles-BV replied:
Added

sidolov commented on app/code/Magento/Wishlist/etc/webapi.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" ?>
+<routes xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Webapi:etc/webapi.xsd">
+    <route method="GET" url="/V1/wishlist/:customerId">
```
Please, use wishlist id

sidolov commented on app/code/Magento/Wishlist/etc/webapi.xml:
```diff
@@ -0,0 +1,15 @@
+    <route method="GET" url="/V1/wishlist/:customerId">
+        <service class="Magento\Wishlist\Api\WishlistRepositoryInterface" method="getWishlistForCustomer"/>
+        <resources>
+            <resource ref="anonymous"/>
```
Please, use separate resources for customer and admin with corresponded access rights. 
Anonymous shouldn't have access to wishlist at all (the only customers may have wishlist)

Tickles-BV replied:
Who should have access? And how do I know who implements the rights? I'm looking at other files for examples but it is confusing. For instance, webapi.xml for lots of modules just use 'Magento_Catalog::products' but I can't find where this is defined. Do you have place with example I can look at?

ishakhsuvarov commented:
Hi @Tickles-BV 
Thank you for this Pull Request. Designing a new API always requires a lot of effort.

I'll try to answer your questions briefly:
> For instance, should you get a list of wishlist products when retrieving a wishlist by customer?

It would be better to handle this in a separate service, getting wishlist contents.

> I've added two methods to the repository should these be in the repository?

Concept of repositories used by Magento only allows to CRUD methods to be present in the `RepositoryInterface`. Methods in question may be moved to a `MagementInterface` or even to a separate commands.

> I've simply made the api route 'V1/wishlist' are there conventions for this? If so, where can I find these?

I'm not aware of any specific convention, so it would be best to make it consistent with other modules and use common sense.

> In the wishlist model, method 'getSharingCode' is added via @property docblock tag, can I add these to api response without adding to new const and wrapping in own method?

I'm pretty sure it would not work with the WebAPI framework. Additionally, it's always the best way to have everything declared explicitly. Wishlist module code seems to be quite "legacy" so it's not a great example.

Please take a look at the code review we have done. It's not 100% complete at the moment, but it doesn't make sense to dig too deep in the code which is subject to change, we could work on it iteratively. 

Thank you for collaboration

Tickles-BV committed:
Rename 'data' namespace to 'Data'

Tickles-BV committed:
Add preference for WishlistInterface

Tickles-BV committed:
Rename current WishlistRepository to WishlistManagement

Tickles-BV committed:
Create new Wishlist Repository Interface with CRUD operations

Tickles-BV committed:
Extend SearchResultsInterface for Wishlist so we can use it for getList

Tickles-BV committed:
Add preference for management interface and searchresultsinterface

Tickles-BV committed:
Re-add wishlist id constant, add setter for customerid

Tickles-BV committed:
Remove typehint WishlistSearchResultsInterface for getList

Tickles-BV committed:
Use api constant for getId function

Tickles-BV committed:
Add CRUD methods to wishlist repository

Tickles-BV committed:
Add functional test for wishlist repository

Tickles-BV committed:
Update test

Tickles-BV committed:
Rename management functions to slightly more descriptive ones

Tickles-BV committed:
Rename get to getById

Tickles-BV committed:
Add repository methods to webapi

Tickles-BV committed:
Add API acl

Tickles-BV committed:
Add copyright header and strict_types declaration

Tickles-BV committed:
Add copyright header and strict_types declaration

Tickles-BV committed:
Add copyright declaration

Tickles-BV committed:
Add copyright declaration

Tickles-BV committed:
Remove unused import

Tickles-BV committed:
Add acl to api and /me variants for self resources

Tickles-BV commented:
Hi @ishakhsuvarov, @sidolov 

I've created WishlistRepository that implements getById, getList, save, delete and deleteById.
Old WishlistRepository methods have been moved to WishlistManagement, but that is probably still wrong.

I've added acl record for Wishlist and added 'me' routes in webapi.xml

Also WishlistRepository is now covered by integration test WishlistRepositoryTest

Please answer my questions about api constants and which getters and setters should be in the interface.

Also should I write WishlistItemRepository in this pull request or is it out of scope?

Thanks.

ishakhsuvarov commented:
@Tickles-BV We are going to review it shortly and get to you with the updates. 

`WishlistItem` interfaces look like they would perfectly fit into the scope of this PR.

Tickles-BV commented:
One more question,

"Setter and getter for extension attributes are missing"

Does this mean I should make Wishlist model extend AbstractExtensibleModel and implement WishlistExtensionInterface and concrete class?

ishakhsuvarov commented on app/code/Magento/Wishlist/Api/Data/WishlistInterface.php:
```diff
@@ -0,0 +1,100 @@
+ */
+namespace Magento\Wishlist\Api\Data;
+
+interface WishlistInterface
```
Please add support for extension attributes to this interface

ishakhsuvarov replied:
Please add type hinting to all method arguments and return types.

ishakhsuvarov replied:
I'd suggest adding wishlist items to this interface as well.

Tickles-BV replied:
I've added function called getItems to wishlist interface (based on naming scheme I could find in other api interfaces) which returns array of wishlist item interfaces

Tickles-BV replied:
I've also made wishlist dto extensible as requested

ishakhsuvarov commented on app/code/Magento/Wishlist/Api/Data/WishlistInterface.php:
```diff
@@ -0,0 +1,100 @@
+
+    const SHARED = 'shared';
+
+    const UPDATED_AT = 'updated_at';
```
There is not need to expose `updated_at` via Wishlist data object. Please consider removing it together with getter and setter.
Possibly it's still valid to maintain this field in `WishlistItem`

Tickles-BV replied:
Removed updated_at property

ishakhsuvarov commented on app/code/Magento/Wishlist/Api/WishlistManagementInterface.php:
```diff
@@ -0,0 +1,27 @@
+
+namespace Magento\Wishlist\Api;
+
+interface WishlistManagementInterface
```
Please add type hints to all methods for arguments and return types

ishakhsuvarov commented on app/code/Magento/Wishlist/Api/WishlistManagementInterface.php:
```diff
@@ -0,0 +1,27 @@
+
+namespace Magento\Wishlist\Api;
+
+interface WishlistManagementInterface
```
Existing wishlist feature supports more actions, e. g. "share wishlist".
There should be investigated and added as well

ishakhsuvarov replied:
I would suggest to split this interface into multiple separate interfaces, pretty much interface per method, as currently, a class which implements this interface (especially the updated version with more features) would have too much responsibilities.

ishakhsuvarov commented on app/code/Magento/Wishlist/Api/WishlistRepositoryInterface.php:
```diff
@@ -0,0 +1,59 @@
+     * @param \Magento\Wishlist\Api\Data\WishlistInterface $wishlist
+     * @return bool
+     */
+    public function delete(\Magento\Wishlist\Api\Data\WishlistInterface $wishlist);
```
It is pretty safe to assume that this method would only proxy to `deleteById`. I would suggest removing it.

Tickles-BV replied:
Deleted

ishakhsuvarov commented on app/code/Magento/Wishlist/Api/WishlistRepositoryInterface.php:
```diff
@@ -0,0 +1,59 @@
+     * @return \Magento\Wishlist\Api\Data\WishlistInterface
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function getById($id): \Magento\Wishlist\Api\Data\WishlistInterface;
```
It would make sense to rename this method to `get`

Tickles-BV replied:
Renamed

ishakhsuvarov commented on app/code/Magento/Wishlist/Api/WishlistManagementInterface.php:
```diff
@@ -0,0 +1,27 @@
+     * @throws \Magento\Framework\Exception\StateException
+     * @return int
+     */
+    public function addWishlistItemByCustomerId($customerId, $sku);
```
Please consider, that wishlist items also support custom options, which can not be resolved via sku only. Possibly it's best to introduce `WishlistItem` and use it here.

ishakhsuvarov commented on app/code/Magento/Wishlist/Api/WishlistRepositoryInterface.php:
```diff
@@ -0,0 +1,59 @@
+     * @return \Magento\Wishlist\Api\Data\WishlistInterface
+     * @throws \Magento\Framework\Exception\CouldNotSaveException
+     */
+    public function save(\Magento\Wishlist\Api\Data\WishlistInterface $wishlist): WishlistInterface;
```
To make this method more CQS-friendly it may be updated to return only id of the created item, instead of the whole item. Returning void may be a good idea as well, but would greatly limit use cases, as there is no predefined if for wishlist.

deefco replied:
Changed to returning id

Tickles-BV replied:
I've changed this method to return just the id

ishakhsuvarov commented on app/code/Magento/Wishlist/Model/Wishlist.php:
```diff
@@ -3,32 +3,30 @@
  * @since 100.0.2
  */
-class Wishlist extends \Magento\Framework\Model\AbstractModel implements \Magento\Framework\DataObject\IdentityInterface
+class Wishlist extends AbstractModel implements IdentityInterface, WishlistInterface
```
This class seems to be a legacy model, and trying to implement `WhishlistInterface` on top of it may be one of the less-optimal solutions. 

I would suggest implementing a separate data transfer object to implement the interface.

Tickles-BV replied:
I've created a new dto in Model\Data\Wishlist which now implements the interface, wishlist now has public function getDataModel which copies properties to new implementation and returns. Repository has been changed to directly return new dto instead of legacy one. All of this has been based on implementation in Customer files Model\Customer, Model\Data\Customer and CustomerRepository.

ishakhsuvarov commented:
Hi @Tickles-BV 
Thank you for the updates. This PR looks much more complete now.

There is still some items missing, main one is the `WishlistItem` and repository for it. Looking more closely at wishlist items it appears that apart from sku it can also bear custom options and qty of the item. It's best to check existing features and make sure interface would cover all of the requirements.

Some features, for example related to sharing, are not covered with this API as well, however I suppose that sharing a wishlist is the main idea of creating it :) Again, it's best to research what's already available, and at least cover all of it with the APIs.

Please let me know if you have any questions regarding the review. You may also contact me and the team via community engineering slack workspace, to speed up communication.

Tickles-BV committed:
Make wishlist extensible

Tickles-BV committed:
Remove updated at calls from interface, add default name so dto doesn't rely on helper

Tickles-BV committed:
Add interface for wishlist items, along with repository

Tickles-BV committed:
Add item repository

Tickles-BV committed:
Remove dto from legacy wishlist class and replace with getDataModel function

Tickles-BV committed:
Create standalone dto for wishlist

Tickles-BV committed:
Add extension interface for wishlist

Tickles-BV committed:
start ItemManagementInterface

Tickles-BV committed:
Add preferences in di.xml

Tickles-BV committed:
Change wishlist repository to return dto instead of legacy object

Tickles-BV committed:
Change getData to _get because of different inheritance

Tickles-BV committed:
Add getItems function to api and wishlist dto

Tickles-BV committed:
Update imports

Tickles-BV committed:
Remove delete because it simply proxied to deleteById

Tickles-BV committed:
Remove delete because it simply proxied to deleteById

Tickles-BV committed:
Remove item management interface because functionality is available in repository getList() function

Tickles-BV committed:
Make save cqrs friendly

Tickles-BV committed:
Prefer the right model for wishlist

Tickles-BV committed:
Use fqcn

Tickles-BV committed:
Make Wishlist Item extensible

Tickles-BV committed:
Legacy model can extend abstractmodel, new one will be extensible

Tickles-BV commented:
I've added wishlist item interface now, along with repository. Repository just implements crud operations same as wishlist repository. I'm going to figure out what other things should be exposed to service contracts.

For example: Sharing wishlist via email appears to be implemented directly in the Shared controller. Should this be extracted to service contract?

Tickles-BV committed:
Add getOptions to Item interface

Tickles-BV committed:
Add search results interface for Repository::getList()

Tickles-BV committed:
Wishlist Item Option DTO now has own Interface (probably bc breaking)

Tickles-BV committed:
Add option search result interface for Repository::getList

Tickles-BV committed:
Add interfaces for item and option repositories

Tickles-BV committed:
Move get back to getById, implement get by sharing code & customer to replicate old api functionality

Tickles-BV committed:
Add strict_types

Tickles-BV committed:
Add first try at SPI interfaces

Tickles-BV committed:
Make option implement own interface

Tickles-BV committed:
Add Concrete Repositories for items and options

Tickles-BV committed:
Add service to add item to wishlist items via interfaces

Tickles-BV committed:
Add service to allow adding option to wishlist item

Tickles-BV committed:
First try at replacing legacy code with spi

Tickles-BV committed:
Allow getting wishlist by sharingcode, id and customer

Tickles-BV committed:
Use right field to load by customer id

Tickles-BV commented:
I've added repository for wishlist item and wishlist item option. Those both now have extended interfaces, item is totally new, option has been changed to new one. Not sure if this effects backwards compatibilty? Option model has had getters and setters changed to adhere to new interface which has type hints, this is definitely b/c breaking, I'd like to hear suggestions on this. Another new DTO perhaps? :)

I'd like opinions on the way the new Wishlist DTO has been implemented. @ishakhsuvarov Said to implement seperate concrete DTO, this has been done by adding getDataModel proxy call that sets data from legacy model onto new model and returns that instead. Implementation is based on what has been done in Customer model (pretty much copied) though someone said this is deprecated way of doing things, it is the only example implementation I could find.

I've added 2 spi classes, AddItemToWishlistInterface and AddOptionToWishlistItemInterface, both based on new MSI model where class has 1 function called 'execute'. I'd like some guidance on figuring out what functionality to extract to SPI layer.

Finally I have a question about DTO implementation. Current Wishlist Model is "clean" implementation (only getters and setters), say I want to have getItems functionality that gets all wishlist items. Should I implement loading these items in plugin on after_load event? Or afterSave()? Or something else entirely? Same goes for saving wishlist, all Items should be saved to keep consistency. Should saving be implemented by plugin? Or afterSave method, or in Repository using SaveHandler? (This is implementation used in Quote Module)

I know these are a lot of questions but I'd like to have a basic idea of where we're going with this pull request because the amount of work keeps growing and I'd really not like to redo any of it because of miscommunication.

Tickles-BV committed:
Add preferences

deefco commented:
@gelanivishal could you please stop changing / doing commits in my pull request? If you require changes to be made please leave them as a comment here so I can discuss this with you. Especially since you are blindly overwriting my changes with core changes while I'm trying to change the core.

gelanivishal commented:
@Tickles-BV please revet back my changes.

misha-kotov commented:
@sidolov @ishakhsuvarov can you guys provide guidance on the above questions by @Tickles-BV ? Let me know if you need me from use case/product perspective.

misha-kotov commented:
I wanted to include a collection of use cases that we need to keep in mind when developing the APIs. We can use these as a sanity check to ensure we have everything in place.
cc: @sidolov @ishakhsuvarov  

Shopper Wishlist Use cases: 
- Wishlist level: 
   - Create wishlist
   - Delete wishlist (can delete all wishlists except id=1)
   - Update wishlist (name, public/private)
   - Share wishlist (list of emails, message)
- Wishlist Item level: 
   - Add wishlist item 
   - Remove wishlist item
   - Update item (configuration, qty, comment)

Use cases we don't need dedicated endpoints for: 
- Move items to wishlist = can be accomplished with `Add wishlist item`, `Remove wishlist item`
- Copy items to wishlist = `Add wishlist item`

In Admin, today we only allow the admin to delete and reconfigure products on customer wishlists. I'm not sure if these are the only ones by design or due to incompletion.

Spaggel commented:
@magento-engcom-team give me test instance

magento-engcom-team commented:
Hi @Spaggel. Thank you for your request. I'm working on Magento instance for you

orlangur reviewed (changes requested):
@Tickles-BV please prepare a new pull request against `2.3-develop` as new features should be only added there: https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#rules
